content/6_Advanced/LCT.mdx

```diff
@@ -30,18 +31,309 @@ frequency: 1
+tree. In one splay operation on $x$, the depths of the nodes on the path from
+the root to $x$ are roughly halved (as seen in the animations below).
+
+<video width="960" height="720" controls>
+	<source src="/animations/splay.mp4" type="video/mp4" />
+</video>
+
+This means that the high cost to access a very deep node is countered by the "work" it does to balance the tree. Unfortunately, this observation alone is not enough to prove the amortized
+$\mathcal{O}(\log n)$ complexity; the full proof remains quite involved.
+
+## Implicit Splays
+
+<FocusProblem problem="dsrars" />
+
+Much like [implicit treaps](https://usaco.guide/adv/treaps#implicit-treaps),
+implicit splays also exist.
+
+### Implementation
+
+**Time Complexity:** $\mathcal{O}((N + Q)\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <functional>
+#include <iostream>
+using ll = long long;
+using namespace std;
+
+// BeginCodeSnip{ModInt class}
+struct mint {
+	const static ll M = 998244353;
+	ll v = 0;
+	mint() {}
+	mint(ll v) { this->v = (v % M + M) % M; }
+	mint operator+(const mint &o) const { return v + o.v; }
+	mint &operator+=(const mint &o) {
+		v = (v + o.v) % M;
+		return *this;
+	}
+	mint operator*(const mint &o) const { return v * o.v; }
+	mint operator-(const mint &o) const { return v - o.v; }
+	mint &operator-=(const mint &o) {
+		mint t = v - o.v;
+		v = t.v;
+		return *this;
+	}
+	friend istream &operator>>(istream &s, mint &v) {
+		s >> v.v;
+		return s;
+	}
+	friend ostream &operator<<(ostream &s, const mint &v) {
+		s << v.v;
+		return s;
+	}
+};
+// EndCodeSnip
+
+struct Line {
+	mint w = 1, b = 0;
+	mint operator()(mint x) { return w * x + b; }
+	Line operator()(Line f) { return Line{w * f.w, w * f.b + b}; }
+	operator bool() const { return w.v != 1 || b.v != 0; }
+};
+
+const int N = 5e5;
+struct Node;
+typedef struct Node *T;
+struct Node {
+	mint v, s;
+	int sz;
+	inline static int n = 0;
+	// Store Nodes in contiguous array rather than allocating with `new` to
+	// prevent MLE
+	static Node t[2 * N];
+	// c[0] -> left child; c[1] -> right child
+	array<T, 2> c{};
+	T p{};
+	Line f{};
+	bool rev = false;
+
+	Node() {}
+	Node(int v) : v(v), s(v), sz(1), rev(false) {}
+
+	static T init(int v, T l = nullptr, T r = nullptr) {
+		t[n] = Node(v);
+		t[n].attach(0, l);
+		t[n].attach(1, r);
+		t[n].pull();
+		return &t[n++];
+	}
+
+	static int size(T x) { return x ? x->sz : 0; }
+
+	static mint sum(T x) { return x ? x->s : 0; }
+
+	void push() {
+		if (rev) {
+			swap(c[0], c[1]);
+			for (T x : c)
+				if (x) { x->rev ^= rev; }
+			rev = false;
+		}
+		if (f) {
+			v = f(v);
+			s = f.w * s + f.b * sz;
+			for (T x : c)
+				if (x) { x->f = f(x->f); }
+			f = Line{1, 0};
+		}
+	}
+
+	void pull() {
+		for (T x : c)
+			if (x) { x->push(); }
+		sz = size(c[0]) + size(c[1]) + 1;
+		s = sum(c[0]) + sum(c[1]) + v;
+	}
+
+	bool side() { return p->c[1] == this; }
+
+	void attach(int i, T x) {
```

Comment:
are users meant to call all of these functions

Reply:
actually these should still be public since they're used in SplayTree (and I'm too lazy to declare friends or whatever)

Reply:
i mean how hard is it to declare friends

Reply:
just realized
could you possibly make the node class like contained within the splay tree class
haven't really taken a look @ the code so i wouldn't know